pysquared/beacon.py

```diff
@@ -86,15 +88,140 @@ def send(self) -> bool:
         Returns:
             True if the beacon was sent successfully, False otherwise.
         """
-        state = self._build_beacon_state()
-        beacon_data = json.dumps(state, separators=(",", ":")).encode("utf-8")
-        return self._packet_manager.send(beacon_data)
+        state = self._build_state()
+        # Use binary encoding for efficiency
+        b = self._encode_binary_state(state)
+        return self._packet_manager.send(b)
+
+    def _encode_binary_state(self, state: OrderedDict[str, object]) -> bytes:
+        """Encode the state dictionary using binary encoding for efficiency.
+
+        Uses explicit encoding based on known beacon data structure for safety and performance.
+
+        Args:
+            state: The state dictionary to encode
+
+        Returns:
+            Binary encoded data
+        """
+        encoder = BinaryEncoder()
+
+        for key, value in state.items():
+            self._encode_known_value(encoder, key, value)
+
+        return encoder.to_bytes()
+
+    def _encode_known_value(
+        self, encoder: BinaryEncoder, key: str, value: object
+    ) -> None:
+        """Encode a value with known type expectations for beacon data.
+
+        This method handles the specific data types we know appear in beacon state,
+        providing explicit and safe encoding without runtime type detection complexity.
+
+        Args:
+            encoder: The binary encoder to add data to
+            key: The key name for the value
+            value: The value to encode
+        """
+        # Handle sensor reading dictionaries (from to_dict() calls)
+        if isinstance(value, dict):
+            self._encode_sensor_dict(encoder, key, value)
+        # Handle lists/tuples - particularly 3D sensor vectors
+        elif isinstance(value, (list, tuple)):
+            if len(value) == 3 and all(isinstance(v, (int, float)) for v in value):
+                # Handle 3D vectors (acceleration, gyroscope) by splitting into components
+                for i, v in enumerate(value):
+                    encoder.add_float(f"{key}_{i}", float(v))
+            else:
+                # Non-numeric or non-3D arrays as strings
+                encoder.add_string(key, str(value))
+        # Handle beacon system data with known types
+        elif key in ("name", "time"):
+            encoder.add_string(key, str(value))
+        elif key == "uptime":
+            encoder.add_float(key, self._safe_float_convert(value))
+        elif "temperature" in key or "voltage" in key or "current" in key:
+            encoder.add_float(key, self._safe_float_convert(value))
+        elif "timestamp" in key:
+            encoder.add_float(key, self._safe_float_convert(value))
```

Comment:
Rather than checking the key for what kind of type the value is, why don't we try just directly checking the type of the value using `isinstance`?
```suggestion
       elif isinstance(value, float)
```